File: r/client/tests/web_socket_test.R
source("client/convex.R")
source("client/live.R")

assert <- function(value, message) {
  if (!isTRUE(value)) stop(message, call. = FALSE)
}

next_port <- local({
  port <- 34000L + (Sys.getpid() %% 10000L)
  function() {
    port <<- port + 1L
    port
  }
})

websocket_accept <- function(key) {
  guid <- "258EAFA5-E914-47DA-95CA-C5AB0DC85B11"
  hash <- digest::digest(charToRaw(paste0(key, guid)), algo = "sha1", serialize = FALSE, raw = TRUE)
  jsonlite::base64_enc(hash)
}

server_frame <- function(opcode, payload, final = TRUE) {
  payload <- if (is.raw(payload)) payload else charToRaw(enc2utf8(payload))
  first <- bitwOr(if (final) 0x80L else 0L, opcode)
  size <- length(payload)
  if (size < 126L) {
    header <- as.raw(c(first, size))
  } else {
    header <- as.raw(c(first, 126L, bitwShiftR(size, 8L), bitwAnd(size, 0xffL)))
  }
  c(header, payload)
}

accept_websocket <- function(port) {
  socket <- socketConnection(
    host = "127.0.0.1",
    port = port,
    server = TRUE,
    blocking = TRUE,
    open = "r+b"
  )
  finish_websocket_handshake(socket)
}

finish_websocket_handshake <- function(socket) {
  request <- readLines(socket, n = 1L, warn = FALSE)
  headers <- list()
  repeat {
    line <- sub("\\r$", "", readLines(socket, n = 1L, warn = FALSE))
    if (!length(line) || !nzchar(line)) break
    pieces <- strsplit(line, ":", fixed = TRUE)[[1L]]
    headers[[tolower(pieces[1L])]] <- trimws(paste(pieces[-1L], collapse = ":"))
  }
  stopifnot(grepl("^GET /api/sync HTTP/1.1", request))
  response <- paste0(
    "HTTP/1.1 101 Switching Protocols\r\n",
    "Upgrade: websocket\r\n",
    "Connection: Upgrade\r\n",
    "Sec-WebSocket-Accept: ", websocket_accept(headers[["sec-websocket-key"]]), "\r\n\r\n"
  )
  writeBin(charToRaw(response), socket)
  flush(socket)
  socket
}

read_exact <- function(socket, size) {
  value <- raw()
  while (length(value) < size) {
    chunk <- readBin(socket, "raw", n = size - length(value))
    if (!length(chunk)) stop("WebSocket peer closed mid-frame")
    value <- c(value, chunk)
  }
  value
}

read_client_frame <- function(socket) {
  header <- as.integer(read_exact(socket, 2L))
  size <- bitwAnd(header[2L], 0x7fL)
  if (size == 126L) {
    extended <- as.integer(read_exact(socket, 2L))
    size <- extended[1L] * 256L + extended[2L]
  } else if (size == 127L) {
    extended <- as.integer(read_exact(socket, 8L))
    if (any(extended[1:4] != 0L)) stop("fixture frame is too large")
    size <- sum(extended[5:8] * c(16777216, 65536, 256, 1))
  }
  masked <- bitwAnd(header[2L], 0x80L) != 0L
  mask <- if (masked) as.integer(read_exact(socket, 4L)) else integer()
  payload <- read_exact(socket, size)
  if (masked && size) {
    payload <- as.raw(bitwXor(as.integer(payload), mask[(seq_len(size) - 1L) %% 4L + 1L]))
  }
  list(opcode = bitwAnd(header[1L], 0x0fL), payload = payload)
}

read_client_json <- function(socket) {
  repeat {
    frame <- read_client_frame(socket)
    if (frame$opcode == 0x1L) {
      return(jsonlite::fromJSON(rawToChar(frame$payload), simplifyVector = FALSE))
    }
  }
}

port <- next_port()
job <- parallel::mcparallel(
  {
    socket <- accept_websocket(port)
    on.exit(close(socket))
    payload <- charToRaw(as.character(convex_json(list(
      type = "Transition",
      startVersion = list(querySet = 0L, identity = 0L, ts = "AAAAAAAAAAA="),
      endVersion = list(querySet = 1L, identity = 0L, ts = "AQAAAAAAAAA="),
      modifications = list(list(
        type = "QueryUpdated",
        queryId = 0L,
        value = list(text = "fragmented 🟨🟩🟦"),
        logLines = list()
      ))
    ))))
    emoji_start <- grepRaw(charToRaw("🟨"), payload, fixed = TRUE)[1L]
    split_at <- emoji_start + 1L
    writeBin(server_frame(0x1L, payload[seq_len(split_at)], final = FALSE), socket)
    # A control frame between fragments proves that transport parsing preserves
    # the message boundary while handling ping/pong.
    writeBin(server_frame(0x9L, charToRaw("ping"), final = TRUE), socket)
    writeBin(server_frame(0x0L, payload[(split_at + 1L):length(payload)], final = TRUE), socket)
    flush(socket)
    Sys.sleep(0.5)
    TRUE
  },
  silent = TRUE
)
Sys.sleep(0.08)
manager <- convex_live(sprintf("http://127.0.0.1:%d", port), "r-websocket-test")
subscription <- manager$subscribe("demo:echo", list())
update <- subscription$next_update(2)
assert(update$value$text == "fragmented 🟨🟩🟦", "fragmented UTF-8 WebSocket message was corrupted")
manager$close()
invisible(suppressWarnings(parallel::mccollect(job, wait = TRUE)))

# Drive five reconnects through six actual TCP/WebSocket handshakes. The peer
# records every Connect and Add, rehydrates with an unchanged value, then sends
# the changed value that should be the only post-disconnect delivery.
port <- next_port()
job <- parallel::mcparallel(
  {
    server <- serverSocket(port)
    on.exit(close(server))
    records <- list()
    for (connection_index in 0:5) {
      socket <- socketAccept(server, blocking = TRUE, open = "r+b")
      socket <- finish_websocket_handshake(socket)
      connect <- read_client_json(socket)
      add <- read_client_json(socket)
      records[[connection_index + 1L]] <- list(connect = connect, add = add)
      start <- list(querySet = 0L, identity = 0L, ts = "AAAAAAAAAAA=")
      hydrated <- list(querySet = 1L, identity = 0L, ts = paste0("real-hydrate-", connection_index))
      hydrated_count <- max(0L, connection_index - 1L)
      writeBin(server_frame(0x1L, as.character(convex_json(list(
        type = "Transition",
        startVersion = start,
        endVersion = hydrated,
        modifications = list(list(type = "QueryUpdated", queryId = 0L, value = list(count = hydrated_count), logLines = list()))
      )))), socket)
      if (connection_index > 0L) {
        changed <- list(querySet = 1L, identity = 0L, ts = paste0("real-change-", connection_index))
        writeBin(server_frame(0x1L, as.character(convex_json(list(
          type = "Transition",
          startVersion = hydrated,
          endVersion = changed,
          modifications = list(list(type = "QueryUpdated", queryId = 0L, value = list(count = connection_index), logLines = list()))
        )))), socket)
      }
      flush(socket)
      repeat {
        frame <- read_client_frame(socket)
        if (frame$opcode == 0x8L) break
      }
      writeBin(server_frame(0x8L, raw()), socket)
      flush(socket)
      close(socket)
    }
    records
  },
  silent = TRUE
)
Sys.sleep(0.08)
manager <- convex_live(sprintf("http://127.0.0.1:%d", port), "r-real-reconnect-test")
subscription <- manager$subscribe("demo:state", list(room = "real-reconnect"))
assert(subscription$next_update(2)$value$count == 0L, "real reconnect fixture lost the initial value")
for (attempt in 1:5) {
  manager$debug_disconnect()
  assert(subscription$next_update(2)$value$count == attempt, "real reconnect relayed hydration or lost the changed value")
}
manager$close()
records <- parallel::mccollect(job, wait = TRUE)[[1L]]
assert(length(records) == 6L, "real reconnect fixture did not open six sockets")
for (connection_index in 0:5) {
  record <- records[[connection_index + 1L]]
  assert(record$connect$type == "Connect", "real reconnect did not send Connect first")
  assert(record$connect$connectionCount == connection_index, "real reconnect lost connectionCount")
  assert(record$add$modifications[[1L]]$type == "Add", "real reconnect did not resend the active Add")
}

# Unsubscribe and close must also remain bounded while the peer continuously
# sends control frames. The library owns frame parsing, while this client keeps
# both operations nonblocking and invalidates delivery before touching I/O.
port <- next_port()
job <- parallel::mcparallel(
  {
    socket <- accept_websocket(port)
    on.exit(close(socket))
    tryCatch(
      repeat {
        writeBin(server_frame(0x9L, charToRaw("busy"), final = TRUE), socket)
        flush(socket)
        Sys.sleep(0.001)
      },
      error = function(error) invisible(NULL)
    )
    TRUE
  },
  silent = TRUE
)
Sys.sleep(0.08)
manager <- convex_live(sprintf("http://127.0.0.1:%d", port), "r-websocket-test")
subscription <- manager$subscribe("demo:state", list(room = "busy-peer"))
manager$pump(0.1)
started <- proc.time()[["elapsed"]]
subscription$close()
assert(proc.time()[["elapsed"]] - started < 0.5, "unsubscribe blocked on a continuously sending peer")
started <- proc.time()[["elapsed"]]
manager$close()
assert(proc.time()[["elapsed"]] - started < 0.5, "close blocked on a continuously sending peer")
tools::pskill(job$pid, tools::SIGTERM)
invisible(suppressWarnings(parallel::mccollect(job, wait = TRUE)))

# Once a partial frame has been consumed, close must not wait for the stalled
# peer or attempt to resume at a false frame boundary.
port <- next_port()
job <- parallel::mcparallel(
  {
    socket <- accept_websocket(port)
    on.exit(close(socket))
    writeBin(as.raw(c(0x81L, 126L, 1L, 0L, 0x7bL)), socket)
    flush(socket)
    Sys.sleep(5)
    TRUE
  },
  silent = TRUE
)
Sys.sleep(0.08)
manager <- convex_live(sprintf("http://127.0.0.1:%d", port), "r-websocket-test")
subscription <- manager$subscribe("demo:state", list(room = "partial-frame"))
manager$pump(0.1)
started <- proc.time()[["elapsed"]]
manager$close()
assert(proc.time()[["elapsed"]] - started < 0.5, "close blocked on a partial WebSocket frame")
tools::pskill(job$pid, tools::SIGTERM)
invisible(suppressWarnings(parallel::mccollect(job, wait = TRUE)))
cat("PASS R WebSocket fragmentation and bounded-close fixtures\n")

File: lua/client/tests/websocket_test.lua
package.path = "./client/?.lua;" .. package.path
local socket = require("socket")
local Live = require("live")

local HOST = "127.0.0.1"
local PORT = 19091

local function frame(fin, opcode, payload)
	payload = payload or ""
	local first = opcode + (fin and 128 or 0)
	if #payload < 126 then
		return string.char(first, #payload) .. payload
	end
	assert(#payload <= 65535, "fixture only emits 16-bit frames")
	return string.char(first, 126, math.floor(#payload / 256), #payload % 256) .. payload
end

local function websocket_handshake(peer)
	local key
	while true do
		local line = assert(peer:receive("*l"))
		if line == "" then
			break
		end
		local name, value = line:match("^([^:]+):%s*(.*)$")
		if name and name:lower() == "sec-websocket-key" then
			key = value
		end
	end
	assert(key, "client handshake omitted Sec-WebSocket-Key")
	local basexx = require("basexx")
	local digest = require("openssl.digest")
	local accept = basexx.to_base64(digest.new("sha1"):final(key .. "258EAFA5-E914-47DA-95CA-C5AB0DC85B11"))
	assert(peer:send("HTTP/1.1 101 Switching Protocols\r\n"))
	assert(peer:send("Upgrade: websocket\r\nConnection: Upgrade\r\n"))
	assert(peer:send("Sec-WebSocket-Accept: " .. accept .. "\r\n\r\n"))
end

local function read_client_frame(peer)
	local header = assert(peer:receive(2))
	local first, second = header:byte(1, 2)
	local length = second % 128
	assert(second >= 128, "client fixture frame must be masked")
	if length == 126 then
		local extended = assert(peer:receive(2))
		local high, low = extended:byte(1, 2)
		length = high * 256 + low
	elseif length == 127 then
		error("fixture does not accept 64-bit client frames")
	end
	local mask = { assert(peer:receive(4)):byte(1, 4) }
	local payload = assert(peer:receive(length))
	local decoded = {}
	local bit = require("http.bit")
	for index = 1, #payload do
		decoded[index] = string.char(bit.bxor(payload:byte(index), mask[(index - 1) % 4 + 1]))
	end
	return first % 16, table.concat(decoded)
end

local function begin_live_subscription(peer)
	local json = require("cjson.safe")
	local _, connect_payload = read_client_frame(peer)
	assert(assert(json.decode(connect_payload)).type == "Connect", "Live client omitted Connect")
	local _, modify_payload = read_client_frame(peer)
	local modify = assert(json.decode(modify_payload))
	assert(modify.type == "ModifyQuerySet", "Live client omitted ModifyQuerySet")
	local add = modify.modifications[1]
	assert(add and add.type == "Add", "Live client omitted Add")
	local transition = assert(json.encode({
		type = "Transition",
		startVersion = { querySet = 0, identity = 0, ts = "AAAAAAAAAAA=" },
		endVersion = { querySet = modify.newVersion, identity = 0, ts = "AQAAAAAAAAA=" },
		modifications = {
			{ type = "QueryUpdated", queryId = add.queryId, value = { count = 0 }, logLines = {} },
		},
	}))
	assert(peer:send(frame(true, 1, transition)))
end

local function run_server()
	local listener = assert(socket.bind(HOST, PORT))
	for scenario = 1, 8 do
		local peer = assert(listener:accept())
		peer:settimeout(2)
		websocket_handshake(peer)
		if scenario == 1 then
			-- Split a UTF-8 codepoint between text fragments and interleave a
			-- control frame. A correct parser validates only the reassembled text.
			local snowman = "\226\152\131"
			assert(peer:send(frame(false, 1, '{"note":"snowman ' .. snowman:sub(1, 1))))
			assert(peer:send(frame(true, 9, "fixture")))
			assert(peer:send(frame(true, 0, snowman:sub(2) .. '"}')))
			local opcode, pong = read_client_frame(peer)
			assert(opcode == 10 and pong == "fixture", "client did not answer fragmented-message ping")
		elseif scenario == 2 then
			-- Time out after consuming a partial extended-length header, then
			-- finish the same frame. Restarting at a false boundary corrupts it.
			assert(peer:send(string.char(129, 126, 0)))
			socket.sleep(0.2)
			assert(peer:send(string.char(126) .. string.rep("p", 126)))
		elseif scenario == 3 then
			-- Stay completely idle while the client initiates close.
			socket.sleep(1)
		elseif scenario == 4 then
			-- Stall after the extended-length marker and one length byte. This
			-- forces a timeout after the parser has already consumed frame bytes.
			assert(peer:send(string.char(129, 126, 0)))
			socket.sleep(1)
		elseif scenario == 5 then
			-- Keep the peer readable while the client closes. A bounded close
			-- must not wait forever for a quiet point in the incoming stream.
			peer:settimeout(0)
			for count = 1, 1000 do
				local ok = peer:send(frame(true, 1, tostring(count)))
				if not ok then
					break
				end
				socket.sleep(0.001)
			end
		else
			begin_live_subscription(peer)
			if scenario == 6 then
				-- Leave the manager's real WebSocket idle during unsubscribe.
				socket.sleep(1)
			elseif scenario == 7 then
				-- Make the owner time out halfway through an extended header while
				-- the controller asks it to remove the active query.
				assert(peer:send(string.char(129, 126, 0)))
				socket.sleep(1)
			else
				-- Keep valid server traffic arriving throughout unsubscribe.
				peer:settimeout(0)
				for _ = 1, 1000 do
					local ok = peer:send(frame(true, 1, '{"type":"Ping"}'))
					if not ok then
						break
					end
					socket.sleep(0.001)
				end
			end
		end
		peer:close()
	end
	listener:close()
end

local function connected_websocket()
	local websocket = require("http.websocket").new_from_uri(string.format("ws://%s:%d/fixture", HOST, PORT))
	assert(websocket:connect(2))
	return websocket
end

local function run_client()
	local cqueues = require("cqueues")
	local cq = cqueues.new()
	cq:wrap(function()
		local fragmented = connected_websocket()
		local payload, kind = fragmented:receive(1)
		assert(kind == "text" and payload == '{"note":"snowman \226\152\131"}', "fragmented UTF-8 changed")
		fragmented:close(1000, "fixture complete", 0.25)

		local partial = connected_websocket()
		local before_timeout = cqueues.monotime()
		local payload_after_partial, _, receive_errno = partial:receive(0.1)
		assert(payload_after_partial == nil and receive_errno, "partial frame did not time out")
		assert(cqueues.monotime() - before_timeout < 0.5, "partial-frame receive ignored its deadline")
		local resumed_payload, resumed_kind = partial:receive(1)
		assert(resumed_kind == "text" and resumed_payload == string.rep("p", 126), "partial frame lost parser state")
		partial:close(1000, "parser fixture", 0.25)

		local idle = connected_websocket()
		local before_idle_close = cqueues.monotime()
		assert(idle:close(1000, "idle fixture", 0.25))
		assert(cqueues.monotime() - before_idle_close < 0.75, "idle-peer close exceeded its deadline")

		partial = connected_websocket()
		payload_after_partial = partial:receive(0.1)
		assert(payload_after_partial == nil, "stalled partial frame unexpectedly completed")
		local before_partial_close = cqueues.monotime()
		assert(partial:close(1000, "partial fixture", 0.25))
		assert(cqueues.monotime() - before_partial_close < 0.75, "partial-frame close exceeded its deadline")

		local continuous = connected_websocket()
		local before_continuous_close = cqueues.monotime()
		assert(continuous:close(1000, "continuous fixture", 0.25))
		assert(cqueues.monotime() - before_continuous_close < 0.75, "continuous-peer close exceeded its deadline")

		for _, peer_state in ipairs({ "idle", "partial", "continuous" }) do
			local manager = Live.Manager.new(string.format("http://%s:%d", HOST, PORT), "lua-test", { cq = cq })
			local subscription = assert(manager:subscribe("demo:state", { room = "unsubscribe-" .. peer_state }))
			assert(assert(subscription:next_update(2)).value.count == 0)
			cqueues.sleep(0.15) -- let the owner enter the hostile receive state
			local before_unsubscribe = cqueues.monotime()
			assert(subscription:close())
			assert(cqueues.monotime() - before_unsubscribe < 0.75, peer_state .. " unsubscribe exceeded its deadline")
			assert(manager:close())
		end
	end)
	assert(cq:loop())
end

if arg[1] == "server" then
	run_server()
elseif arg[1] == "client" then
	run_client()
else
	error("expected server or client mode")
end
